validator/sawtooth_validator/journal/completer.py

```diff
@@ -36,93 +39,266 @@ class Completer(object):
+            return
 
+        # used to supplement batch_cache, contains batches already in block
+        temp_batches = {}
+        for batch in block.batches:
+            temp_batches[batch.header_signature] = batch
+
+        # The block is missing batches. Check to see if we can complete it.
         if len(block.batches) != len(block.header.batch_ids):
-            LOGGER.debug("Block discarded(Missing batches): %s",
-                         block.header_signature[:8])
-            valid = False
-
-        for i in range(len(block.header.batch_ids)):
-            if block.batches[i].header_signature != block.header.batch_ids[i]:
-                LOGGER.debug("Block discarded(Missing batch): %s",
-                             block.header_signature[:8])
-                if block.header.batch_ids[i] not in self.batch_cache:
-                    LOGGER.debug("Request missing batch: %s",
-                                 block.header.batch_ids[i])
-                    self.gossip.broadcast_batch_by_batch_id_request(
-                        block.header.batch_ids[i])
-                valid = False
+            building = True
+            for batch_id in block.header.batch_ids:
+                if batch_id not in self.batch_cache and \
+                        batch_id not in temp_batches:
+                    # Request all missing batches
+                    self.gossip.broadcast_batch_by_batch_id_request(batch_id)
+                    if batch_id not in self._incomplete_blocks:
+                        self._incomplete_blocks[batch_id] = [block]
+                    elif block not in self._incomplete_blocks[batch_id]:
+                        self._incomplete_blocks[batch_id] += [block]
+                    building = False
+
+            if not building:
+                # The block cannot be completed.
+                return
+
+            batches = self._finalize_batch_list(block, temp_batches)
+            del block.batches[:]
+            # reset batches with full list batches
+            block.batches.extend(batches)
+            return block
+
+        else:
+            batch_id_list = [x.header_signature for x in block.batches]
+            # Check to see if batchs are in the correct order.
+            if batch_id_list == list(block.header.batch_ids):
+                return block
+            # Check to see if the block has all batch_ids and they can be put
+            # in the correct order
+            elif sorted(batch_id_list) == sorted(list(block.header.batch_ids)):
+                batches = self._finalize_batch_list(block, temp_batches)
+                # Clear batches from block
+                del block.batches[:]
+                # reset batches with full list batches
+                if batches is None:
+                    block.batches.extend(batches)
+                return block
+            else:
+                LOGGER.debug("Block has extra batches. Dropping %s", block)
```

Comment:
Perhaps "block.header.batch_ids does not match set of batches in block.batches" ?